### ISSUE TITLE
Offer a "Move Brim to Current Display" option

### ISSUE THREAD

**Issue by philrz** [brimsec/brim#1041]
@orochford reported that when he removed his laptop from two external displays and then tried to open Brim, and it's no longer visible. He pointed out that apps like Slack offer a right-click option that can repair this, which leads us to guess that maybe this is a common problem with Electron apps.

![image](https://user-images.githubusercontent.com/5934157/91995976-51ae4080-eced-11ea-90d6-174190df2e42.png)

In past situations when this has happened to me with other Windows apps I'd had success hitting Alt-Tab repeatedly until I was effectively "on" the window of the app I couldn't see, then hit Windows-LeftArrow or Windows-RightArrow enough times to step it onto my visible desktop. However, @orochford reports that did not work for him, so I've pointed him at the [Troubleshooting article for clearing app state](https://github.com/brimsec/brim/wiki/Troubleshooting#brim-seems-unable-to-restart-normally-such-as-after-a-bad-crash) to get him unstuck.

**Comment by philrz** [comment]
I've not been successful in my attempts to verify this. I imagine there might be multiple ways to get into a state where this enhancement is needed, so I'll lay out the details of what I did and what I saw and we can compare experiences.

In my setup I have a Windows laptop and an external monitor. Using GA Brim tagged `v0.18.0`, I found I wasn't able to repro the issue "removing" my laptop from the display (that is, yanking the cable that goes from my laptop to the monitor). I found that if I had Brim open on the monitor before doing this, Windows simply moved the Brim window to the remaining laptop display _for me_ without any action required. Then when I restored the connection and the laptop saw two available displays again, it automatically moved the Brim window back to the external display. Therefore everything seemed to be working fine.

How I _was_ able to find a need for this feature, though, was by leaving the external monitor connected, but not accessible to me as a user. My monitor has a push-button feature to toggle between showing the output from one of two devices, and in my case device 1 is the Windows laptop and device 2 is my Mac laptop. So now I open Brim on the external Windows desktop, then toggle the monitor so it's showing the Mac desktop. Now the task bar on the Windows laptop display shows the Brim app as being open, but when I click on that task bar entry, there's no option to move it to the current display. If I right-click and select options like "New Window" or "Brim", it just opens more Brim windows on the external display that's active-but-I-can't-see-it. This seems like a legitimate starting point for validating the issue. The attached video captures the user experience of struggling with the task bar icon while on the laptop display.

[Repro.zip](https://github.com/brimsec/brim/files/5409652/Repro.zip)

To attempt my verification, I created and installed a release artifact at Brim commit `413f9ad`, which adds the right-click option to "Move Brim to Current Display". Starting from the same repro steps described above, this next video shows what's happening on the external display while I'm selecting that option on the laptop display. As you can see, the Brim window _does_ get moved, but it ultimately remains on the external display, hence still out of reach for me as a user.

[Verify-Attempt-1.zip](https://github.com/brimsec/brim/files/5409662/Verify-Attempt-1.zip)

I actually saw a slightly different variation in my first attempt last night, but I don't have video or screenshot of it. In that case it moved the Brim window such that it was partially on the external display and partially on the laptop display. However, the very top of the Brim window was "just off the top" of the laptop display such that I could not click on it to move it into the main area of the laptop display. So, once again, it did not seem to address the problem.

Based on these observations, I can only assume that this functionality must be sensitive to the precise size and positioning of the window's starting position and perhaps the relative size and orientation of the multiple displays. Is it possible we need to simplify some logic to make it more deterministic? I don't claim an encyclopedic knowledge of how Slack handles this same problem, but in my minimal tinkering it looks as if their "Move Slack to Current Display" option not only moves it but _maximizes_ it, perhaps in some brute force attempt to make sure that it's fully "there". I know we have the potential to have multiple Brim windows open at once, so maybe we can't employ the same simple approach. But based on my experience it does look like we have more left to do here to make this enhancement bulletproof.

**Comment by philrz** [comment]
Now with the benefit of the additional fix from #1158, I can verify that things are much improved, but have discovered one more loose end.

Because of the multi-display action, this one doesn't lend itself to the usual proof-by-video treatment, so I'll use some before/after screenshots.

Starting with a completely fresh environment, I have Brim open on the external display (left) while the built-in laptop display is blank (right).

![image](https://user-images.githubusercontent.com/5934157/96798586-0f80b180-13b6-11eb-9705-1c5b729b2673.png)

Right-clicking the Brim icon in the task bar on the built-in laptop display and selecting "Move Brim to Current Display", it goes fully into that display and is completely inside the window.

![image](https://user-images.githubusercontent.com/5934157/96798673-3a6b0580-13b6-11eb-85e6-0f338d539093.png)

If I do the same right-click now in the external display, it moves back.

![image](https://user-images.githubusercontent.com/5934157/96798700-4656c780-13b6-11eb-8663-170e5aca8830.png)

When testing at a previous revision, I had bad results with smaller windows that started at different spots in the external display, but these work fine now.

Here's a similar trio of screenshots with a smaller Brim window starting in the upper-left, working fine:

![image](https://user-images.githubusercontent.com/5934157/96798872-aa798b80-13b6-11eb-829d-7f3459f5f845.png)

![image](https://user-images.githubusercontent.com/5934157/96798898-b7967a80-13b6-11eb-92dd-262884554a30.png)

![image](https://user-images.githubusercontent.com/5934157/96798914-bf561f00-13b6-11eb-9591-5af5cb700e2e.png)

Here's a similar trio of screenshots with a smaller Brim window starting in the lower-right, also working fine:

![image](https://user-images.githubusercontent.com/5934157/96798947-db59c080-13b6-11eb-9620-f55ffaf9be1a.png)

![image](https://user-images.githubusercontent.com/5934157/96798970-e7de1900-13b6-11eb-984c-3d8a84e92465.png)

![image](https://user-images.githubusercontent.com/5934157/96798990-f2001780-13b6-11eb-8761-bb956b70922a.png)

Unfortunately it looks like we still have a remaining corner case involving maximized windows. If I start out with Brim maximized on the smaller built-in laptop display, I _can_ move it successfully to the larger external display.

![image](https://user-images.githubusercontent.com/5934157/96799053-21af1f80-13b7-11eb-983f-5fd00bafd228.png)

![image](https://user-images.githubusercontent.com/5934157/96799061-283d9700-13b7-11eb-9db9-ee696d6d328a.png)

It also works nicely if I've got multiple windows open.

![image](https://user-images.githubusercontent.com/5934157/96799466-07c20c80-13b8-11eb-9475-af3a520adcf9.png)

![image](https://user-images.githubusercontent.com/5934157/96799486-13153800-13b8-11eb-9862-bcfbf3a05c5d.png)

![image](https://user-images.githubusercontent.com/5934157/96799503-1b6d7300-13b8-11eb-9b65-1f9370a532a9.png)

However if I start out with the Brim window maximized in the larger, external display:

![image](https://user-images.githubusercontent.com/5934157/96798730-566ea700-13b6-11eb-802b-2bd8acaea019.png)

...I find I can't move it to the smaller, built-in laptop display. When I right-click, it effectively acts like a silent no-op. There's no complaints in DevTools Console either.

Re-opening. The window-moving beatings shall continue until window-moving morale improves! 😄

**Comment by philrz** [comment]
Verified in Brim commit `4da1755`.

First I repeated the verifications described in the previous comment to make sure they were still ok, and those all verified fine. Then for the case where the Brim window starts out maximized on the larger external display, it now moves to the smaller laptop display just fine.

![image](https://user-images.githubusercontent.com/5934157/97024119-4ae0c480-150b-11eb-8d8a-decb1a6525cc.png)

![image](https://user-images.githubusercontent.com/5934157/97024171-5502c300-150b-11eb-91f2-41200c575417.png)

![image](https://user-images.githubusercontent.com/5934157/97024202-5e8c2b00-150b-11eb-8819-8669dfa235f2.png)

Thanks @jameskerr!
